srml/support/src/storage/generator/linked_map.rs

```diff
@@ -47,31 +47,59 @@ pub trait StorageLinkedMap<K: FullCodec, V: FullCodec> {
 	/// Hasher used to insert into storage.
 	type Hasher: StorageHasher;
 
+	/// The family of key formats used for this map.
+	type KeyFormat: KeyFormat<Hasher=Self::Hasher>;
+
 	/// Prefix used to prepend each key.
 	fn prefix() -> &'static [u8];
 
-	/// Key used to store linked map head.
-	fn head_key() -> &'static [u8];
+	/// The head key of the linked-map.
+	fn head_key() -> &'static [u8] {
+		<Self::KeyFormat as KeyFormat>::head_key()
+	}
 
-	/// Convert an optionnal value retrieved from storage to the type queried.
+	/// Convert an optional value retrieved from storage to the type queried.
 	fn from_optional_value_to_query(v: Option<V>) -> Self::Query;
 
-	/// Convert a query to an optionnal value into storage.
+	/// Convert a query to an optional value into storage.
 	fn from_query_to_optional_value(v: Self::Query) -> Option<V>;
 
 	/// Generate the full key used in top storage.
 	fn storage_linked_map_final_key<KeyArg>(key: KeyArg) -> <Self::Hasher as StorageHasher>::Output
 	where
 		KeyArg: EncodeLike<K>,
 	{
-		let mut final_key = Self::prefix().to_vec();
-		key.encode_to(&mut final_key);
-		Self::Hasher::hash(&final_key)
+		<Self::KeyFormat as KeyFormat>::storage_linked_map_final_key::<KeyArg>(Self::prefix(), &key)
 	}
 
 	/// Generate the hashed key for head
 	fn storage_linked_map_final_head_key() -> <Self::Hasher as StorageHasher>::Output {
-		Self::Hasher::hash(Self::head_key())
+		<Self::KeyFormat as KeyFormat>::storage_linked_map_final_head_key()
+	}
+}
+
+/// A type-abstracted key format used for a family of linked-map types.
+pub trait KeyFormat {
```

Comment:
> I mostly made a bunch of stuff more permissive so that we could use the definition of a StorageMap<K, V> to also look up keys for a StorageMap<K2, V2>, which is necessary for translation. Most of that hinges around an internal KeyFormat trait that every StorageMap has as an assoc. type, which is un-opinionated about the actual types passed to it.

I'm not sure I understand why you introduce this trait ? to access the old value before the translation ? because in this PR it seems the whole code (both translation implementation and the test) doesn't make use of this KeyFormat extraction.

Reply:
it is used, heavily. 

> to access the old value before the translation ?

yes, exactly.

Reply:
why would you want to access it before the translation ? isn't translation the first and only call you need to do ?

Reply:
Yes, translation is the first and only call you need to do. `KeyFormat` is used within the body of the `translate` function to access the old values before the translation has actually occurred - that's how I interpreted "accessing" the old value before translations.

Reply:
oh ok then I would have just introduced two private function
```rust
fn read_head<Key, Value>(key: key, prefix: &'static [u8]) -> Value {...}
fn final_key<Key>(key: key, prefix: &'static [u8]) -> Vec<u8> {...}
```
instead of the new trait but ok.

Reply:
Including those as free functions (i.e. fixing logic for key generation) defeats the purpose of having a trait function for computing the final keys in the first place. The point of having `fn storage_linked_map_final_key` is so that the implementation can decide how to compute the key. 

All of the changes I included are a refactoring - the code before this diff seriously misunderstood Rust traits.